src/main/java/org/dependencytrack/policy/cel/CelPolicyLibrary.java

```diff
@@ -143,11 +157,84 @@ public List<ProgramOption> getProgramOptions() {
+        }
+        if (!(vals[2].value() instanceof final String comparator)) {
+            return Err.maybeNoSuchOverloadErr(vals[2]);
+        }
+        if (!component.hasPurl()) {
+            return Err.newErr("Provided component does not have a purl", vals[0]);
+        }
+        try {
+            if (RepositoryType.resolve(new PackageURL(component.getPurl())) == RepositoryType.UNSUPPORTED) {
+                return Err.newErr("Unsupported repository type for component: ", vals[0]);
+            }
+        } catch (MalformedPackageURLException ex) {
+            return Err.newErr("Invalid package url ", component.getPurl());
+        }
+        if (!component.hasLatestVersion()) {
+            return Err.newErr("Requested component does not have latest version information", component);
+        }
+        return Types.boolOf(matchesVersionDIstance(component, value, comparator));
+    }
+
+    private static boolean matchesVersionDIstance(Component component, String value, String comparator) {
+        String comparatorComputed = switch (comparator) {
+            case "NUMERIC_GREATER_THAN", ">" -> "NUMERIC_GREATER_THAN";
+            case "NUMERIC_GREATER_THAN_OR_EQUAL", ">=" -> "NUMERIC_GREATER_THAN_OR_EQUAL";
+            case "NUMERIC_EQUAL", "==" -> "NUMERIC_EQUAL";
+            case "NUMERIC_NOT_EQUAL", "!=" -> "NUMERIC_NOT_EQUAL";
+            case "NUMERIC_LESSER_THAN_OR_EQUAL", "<=" -> "NUMERIC_LESSER_THAN_OR_EQUAL";
+            case "NUMERIC_LESS_THAN", "<" -> "NUMERIC_LESS_THAN";
+            default -> "";
+        };
+        if (comparatorComputed.equals("")) {
+            LOGGER.warn("Was passed a not supported operator for version distance policy " + comparator);
+            return false;
+        }
+        final VersionDistance versionDistance;
+        try {
+            versionDistance = VersionDistance.getVersionDistance(component.getVersion(), component.getLatestVersion());
+        } catch (RuntimeException e) {
+            LOGGER.warn("""
+                    Failed to compute version distance for component %s (UUID: %s), \
+                    between component version %s and latest version %s; Skipping\
+                    """.formatted(component, component.getUuid(), component.getVersion(), component.getLatestVersion(), e));
+            return false;
+        }
+        return isDirectDependency(component) && VersionDistance.evaluate(value, comparatorComputed, versionDistance);
+    }
+
+    static boolean isDirectDependency(final org.dependencytrack.proto.policy.v1.Component component) {
+        try (QueryManager queryManager = new QueryManager()) {
+            org.dependencytrack.model.Component componentFromDb = queryManager.getObjectByUuid(org.dependencytrack.model.Component.class, UUID.fromString(component.getUuid()));
+            return componentFromDb.getProject().getDirectDependencies().contains("\"uuid\":\"" + component.getUuid() + "\"");
```

Comment:
This check should be done in a single query, without fetching the component and project from the database.